### ISSUE TITLE
[rllib] Dummy inputs (all zeros!) still often exist in the train_batch when training ppo after initializing the loss

### ISSUE THREAD

**Issue by yangysc** [ray-project/ray#13183]
### What is the problem?
When the policy creates the self._dummy_batch using [self._get_dummy_batch_from_view_requirements(sample_batch_size)](https://github.com/ray-project/ray/blob/master/rllib/policy/policy.py#L615-L616), the dummy inputs still exist in the sampler workers. And the function [ppo_surrogate_loss](https://github.com/ray-project/ray/blob/master/rllib/agents/ppo/ppo_torch_policy.py#L47) often uses the dummy inputs to train the agent. 

Should we delete them  after initializing the loss? What is the point of training the agent sometimes using all zero inputs? And this could be very bad for autoregressive action dist, since all actions are zero, and I think it can heavily influence the performance of the agent (like it can change the gradient). How can we remove them? Any suggestion? @sven1977   @ericl   Thanks a lot~

![train_batch['obs'].numpy()](https://user-images.githubusercontent.com/8980981/103611925-4b132c80-4f5e-11eb-8a24-a20035c09f0d.png)

![train_batch['action_dist_inputs'].numpy()](https://user-images.githubusercontent.com/8980981/103611929-4f3f4a00-4f5e-11eb-9a24-0f6299b4f006.png)

*Ray version and other system information (Python version, TensorFlow version, OS):*
- Ray: 1.2.0.dev.0; 
- Python 3.8
- Pytorch 1.7.1
- TF: 2.3.0
- OS: Ubuntu 20.04

### Reproduction (REQUIRED)
Run the official example of [cartpole_lstm.py](https://github.com/ray-project/ray/blob/master/rllib/examples/cartpole_lstm.py) in the local_mode with Pytorch, after training one iteration, we can set a breakpoint in the function [ppo_surrogate_loss](https://github.com/ray-project/ray/blob/master/rllib/agents/ppo/ppo_torch_policy.py#L47), and we can see there are some zero inputs in **train_batch**

If the code snippet cannot be run by itself, the issue will be closed with "needs-repro-script".

- [x] I have verified my script runs in a clean environment and reproduces the issue.
- [x] I have verified the issue also occurs with the [latest wheels](https://docs.ray.io/en/master/installation.html).

**Comment by ericl** [comment]
This might just be normal zero padding for the lstm sequences (they will be masked out of the loss), but cc @sven1977 to confirm

**Comment by yangysc** [comment]
> This might just be normal zero padding for the lstm sequences (they will be masked out of the loss), but cc @sven1977 to confirm

Thanks! It seems to be zero padding for lstm, as you said, and they are indeed masked out of the loss. But calling model.from_batch (Line 47, ppo_torch_policy.py) on these zero paddings means the policy network still forwards the zero padding input. Should the zero padding be masked out **before the from_batch call, instead of before the loss?** It could be more efficient, if the policy network is time-consuming. For example, for the cartpole_lstm.py example, the num_valid obs (without zeros padding) is 128, and the size of the total obs (including zeros padding) is 160, meaning 32/160=0.2 fraction of the operation is meaningless.

**Comment by ericl** [comment]
Indeed we should mask them out there as well. There was an old PR that did
that, but there were some complications around mini batching (it became
difficult to deal with mini batching for PPO without the padding).

Probably it could be solved with some smarter handling, but this was never
fixed so we have the padding for now (and the associated performance cost).
